common/src/main/java/com/revenuecat/purchases/common/Factories.kt

```diff
@@ -130,15 +130,22 @@ private fun JSONObject.buildEntitlementInfo(
     val expirationDate = optDate("expires_date")
     val unsubscribeDetectedAt = productData.optDate("unsubscribe_detected_at")
     val billingIssueDetectedAt = productData.optDate("billing_issues_detected_at")
+
+    val store = productData.getStore("store")
+    val isPromo = store == Store.PROMOTIONAL
+    val isLifetime = expirationDate == null
+    val hasUnsubscribed = unsubscribeDetectedAt != null
+    val hasBillingIssues = billingIssueDetectedAt != null
+
     return EntitlementInfo(
         identifier = identifier,
         isActive = expirationDate == null || expirationDate.after(requestDate ?: Date()),
-        willRenew = expirationDate == null || (unsubscribeDetectedAt == null && billingIssueDetectedAt == null),
+        willRenew = !(isPromo || isLifetime || hasUnsubscribed || hasBillingIssues),
```

Comment:
kinda feels like this really should be a nullable field, but I guess if the API is already returning bool we should do the same

Reply:
is this worth extracting as a method?

Reply:
Yeah good call on the method